### PR TITLE
SNOW-1305528 : Supporting case sensitive objects in session.lineage.trace

### PR DESCRIPTION
**Pull request by sfc-gh-rsureshbabu**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   The PR adds support for handling case sensitive name of the objects. 
   1. Constructs case sensitive names and versions 
   2. Queries DGQL by Id for distance >= 2. This also helps in going beyond deleted objects.  

3. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

4. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

### CODE REVIEW

File: src/snowflake/snowpark/lineage.py [hunk 1]
```diff
@@ -116,7 +121,7 @@ class _UserDomain:
     """
 
     FEATURE_VIEW = "FEATURE_VIEW"
-    MODEL = "MODULE"
+    MODEL = "MODEL"
```
[the remark this answers is not in the document]
Comment (sfc-gh-rsureshbabu) [reply src/snowflake/snowpark/lineage.py:124]: yep.

File: src/snowflake/snowpark/lineage.py [hunk 2]
```diff
@@ -182,64 +186,121 @@ def _build_graphql_query(
...
+                        f"Version cant be empty for {_UserDomain.FEATURE_VIEW}"
+                    )
+                object_name = _DGQLQueryBuilder._get_feature_view_name(
+                    object_name, object_version
+                )
+                object_version = None
+            object_name = object_name.replace('"', '\\\\"')
+            query_object = object_name
+            if object_version:
+                object_version = object_version.replace('"', '\\\\"')
```
[the remark this answers is not in the document]
Comment (sfc-gh-rsureshbabu) [reply src/snowflake/snowpark/lineage.py:209]: Model version is SQL identifier but dataset version is a case sensitive string.

File: src/snowflake/snowpark/lineage.py [hunk 3]
```diff
@@ -182,64 +186,121 @@ def _build_graphql_query(
...
-
         return f"select SYSTEM$DGQL('{query}')"
 
+    @staticmethod
+    def split_fully_qualified_name(name):
+        """
+        Splits the fully qualified name.
+        Pattern matches either a string enclosed in double quotes or a sequence of word characters.
+        """
+        parts = re.findall(r'"[^"]*"|\w+', name)
```
[the remark this answers is not in the document]
Comment (sfc-gh-rsureshbabu) [reply src/snowflake/snowpark/lineage.py:234]: \w+ matches for word characters (alpha numeric and underscore), dot(.) is not a underscore.

File: src/snowflake/snowpark/lineage.py [hunk 4]
```diff
@@ -182,64 +186,121 @@ def _build_graphql_query(
...
+            if object_domain == _UserDomain.FEATURE_VIEW:
+                if not object_version:
+                    raise ValueError(
+                        f"Version cant be empty for {_UserDomain.FEATURE_VIEW}"
+                    )
+                object_name = _DGQLQueryBuilder._get_feature_view_name(
+                    object_name, object_version
+                )
+                object_version = None
+            object_name = object_name.replace('"', '\\\\"')
```
[the remark this answers is not in the document]
Comment (sfc-gh-rsureshbabu) [reply src/snowflake/snowpark/lineage.py:206]: yes

File: src/snowflake/snowpark/lineage.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sfc-gh-yli) [reply src/snowflake/snowpark/lineage.py:209]: ok, so then let me also add handling for the versions when returning from DGQL

File: tests/integ/test_lineage.py [hunk 5]
```diff
@@ -185,26 +185,59 @@ def test_lineage_trace(session):
...
+        "DISTANCE": [1, 2],
     }
 
     expected_df = pd.DataFrame(expected_data)
     assert_frame_equal(df, expected_df, check_dtype=False)
 
-    session.sql(f"drop schema {db}.{schema}").collect()
+    # CASE 6 : Case senstive query
+    session.sql(
+        f'CREATE OR REPLACE VIEW {db}.{schema}."v7" AS SELECT * FROM {db}.{schema}.V2'
```
[the remark this answers is not in the document]
Comment (sfc-gh-rsureshbabu) [reply tests/integ/test_lineage.py:213]: yes. I have a unit test that ensures we support parsing that cases correctly.

File: src/snowflake/snowpark/lineage.py [hunk 6]
```diff
@@ -182,64 +186,121 @@ def _build_graphql_query(
...
+        """
+        parts = _DGQLQueryBuilder.split_fully_qualified_name(name)
+        if len(parts) < 3:
+            raise ValueError("Invalid object name: less than three parts")
+
+        # Unquote the 3rd part if it exists
+        name_part = parts[2].strip('"')
+
+        # Feature view name is case insensitive
+        feature_view_name = f"{name_part.upper()}${version}"
```
Comment (sfc-gh-yli) [inline src/snowflake/snowpark/lineage.py:250]: is this true? customers cannot name a DT/feature view with quoted names in lower case?

File: tests/unit/test_lineage.py [hunk 7]
```diff
@@ -175,3 +177,132 @@ def test_get_user_entity():
...
+        Lineage(fake_session)._check_valid_object_name(case)
+
+    # Assert checks for invalid cases
+    for case in test_cases_invalid:
+        with pytest.raises(ValueError) as exc:
+            Lineage(fake_session)._check_valid_object_name(case)
+        assert "Invalid object name:" in str(exc)
+
+
+def test_get_feature_view_name():
```
Comment (sfc-gh-yli) [inline tests/unit/test_lineage.py:225]: can we test cases where the database/schema names are quoted and have special characters such as $ or dot?

File: src/snowflake/snowpark/lineage.py (hunk 6 above)
Comment (sfc-gh-rsureshbabu) [reply src/snowflake/snowpark/lineage.py:250]: The feature view name is case insensitive. But the version is sensitive. To treat the dynamic table/view name which is a combination of both as case sensitive, we always keep the name is upper case.

File: tests/unit/test_lineage.py (hunk 7 above)
Comment (sfc-gh-rsureshbabu) [reply tests/unit/test_lineage.py:225]: added

File: src/snowflake/snowpark/lineage.py (hunk 6 above)
Comment (sfc-gh-rsureshbabu) [reply src/snowflake/snowpark/lineage.py]: > customers cannot name a DT/feature view with quoted names in lower case?
I did a test, it turns out that they can. I have changed the logic.

**Review by sfc-gh-yli — APPROVED** [review]
LGTM!

**Review by sfc-gh-yixie — APPROVED** [review]

**Pull request merged**